gadgets/trace_dns/program.bpf.c

```diff
@@ -269,7 +269,27 @@ int ig_trace_dns(struct __sk_buff *skb)
 		dport = load_half(skb, l4_off + offsetof(struct udphdr, dest));
 		dns_off = l4_off + sizeof(struct udphdr);
 		break;
-	// TODO: support TCP
+	case IPPROTO_TCP:
+		// This is best effort, since we don't reassemble segments.
+		struct tcphdr tcph;
+		if (bpf_skb_load_bytes(skb, l4_off, &tcph, sizeof tcph))
+			return 0;
```

Comment:
Here we copy 20 bytes instead of just the ports. This means a bit more work to do on all tcp traffic, including on traffic unrelated to DNS.

Also, by luck, we have:
- `offsetof(struct udphdr, source)` == `offsetof(struct tcphdr, source)`
- `offsetof(struct udphdr, dest)` == `offsetof(struct tcphdr, dest)`

So we could do the `is_dns_port` test before the `switch (proto)`.

I don't know if the optimization is worth it, wdty?

Reply:
Do mean something following?

```
    // Since we have the same offset for both UDP and TCP, we can use the same code to extract the ports for both.
    // - offsetof(struct udphdr, source) == offsetof(struct tcphdr, source)
    // - offsetof(struct udphdr, dest) == offsetof(struct tcphdr, dest)
	switch (proto) {
	case IPPROTO_UDP:
	case IPPROTO_TCP:
		sport = load_half(skb,
				  l4_off + offsetof(struct udphdr, source));
		dport = load_half(skb, l4_off + offsetof(struct udphdr, dest));
		break;
	default:
		return 0;
	}

	if (!is_dns_port(sport) && !is_dns_port(dport))
		return 0;
```

It works fine. Also, I feel it always better to return early.